### PR TITLE
Proxy environment variables not being set correctly for authenticated proxies

### PR DESCRIPTION
**Pull request by benPearce1**
Fixes: https://github.com/OctopusDeploy/Issues/issues/4900

### CODE REVIEW

File: source/Calamari.Tests/Fixtures/PowerShell/PowerShellFixture.cs [hunk 1]
```diff
@@ -522,5 +523,123 @@ public void ShouldAllowPlatformSpecificScriptToExecute()
...
+            output.AssertSuccess();
+            output.AssertOutputContains($"HTTP_PROXY: {httpProxy}");
+            output.AssertOutputContains($"HTTPS_PROXY: {httpsProxy}");
+
+            ResetProxyEnvironmentVariables();
+        }
+
+        [Test]
+        [Category(TestEnvironment.CompatibleOS.Windows)]
+        public void ProxySetToSystem_ShouldSetSetVariablesCorrectly()
```
Comment (johnsimons) [inline source/Calamari.Tests/Fixtures/PowerShell/PowerShellFixture.cs]: Spelling mistake, "Set" twice

File: source/Calamari.Shared/Integration/Scripting/WindowsPowerShell/Bootstrap.ps1 [hunk 2]
```diff
@@ -403,27 +436,16 @@ function Initialize-ProxySettings()
 	[int]$proxyPort = $env:TentacleProxyPort
 	
 	$useSystemProxy = [string]::IsNullOrEmpty($proxyHost) 
-	
 	if($useSystemProxy)
 	{
 		$proxy = [System.Net.WebRequest]::GetSystemWebProxy()
+		Set-ProxyEnvironmentVariablesFromSystemProxy -proxyUsername $proxyUsername -proxyPassword $proxyPassword
```
Comment (johnsimons) [inline source/Calamari.Shared/Integration/Scripting/WindowsPowerShell/Bootstrap.ps1]: There are still formatting issues with tabs vs spaces

File: source/Calamari.Shared/Integration/Scripting/WindowsPowerShell/Bootstrap.ps1 [hunk 3]
```diff
@@ -395,6 +395,39 @@ function Decrypt-Variables($iv, $Encrypted)
...
+function Set-ProxyEnvironmentVariablesFromSystemProxy([string] $proxyUsername, [string] $proxyPassword) {
+	$testUri = New-Object Uri("https://octopus.com")
+	$systemProxy = [System.Net.WebRequest]::GetSystemWebProxy().GetProxy($testUri)
+	if ($systemProxy.Host -ne "octopus.com") {
+		Set-ProxyEnvironmentVariables -proxyHost $systemProxy.Host -proxyPort $systemProxy.Port -proxyUsername $proxyUsername -proxyPassword $proxyPassword
+	}
+}
+
+function Get-ProxyUri ([string] $proxyHost, [int] $proxyPort) {
+	$uri = "http://${proxyHost}:$proxyPort"
```
Comment (johnsimons) [inline source/Calamari.Shared/Integration/Scripting/WindowsPowerShell/Bootstrap.ps1]: Shouldn't it be `https` for `$env:HTTPS_PROXY` ?

File: source/Calamari.Shared/Integration/Scripting/WindowsPowerShell/Bootstrap.ps1 [hunk 4]
```diff
@@ -395,6 +395,39 @@ function Decrypt-Variables($iv, $Encrypted)
 	return $parameters
 }
 
+function Set-ProxyEnvironmentVariables ([string] $proxyHost, [int] $proxyPort, [string] $proxyUsername, [string] $proxyPassword) {
+    $proxyUri = Get-ProxyUri -proxyHost $proxyHost -proxyPort $proxyPort
+    if (![string]::IsNullOrEmpty($proxyUsername)) {
+		Add-Type -AssemblyName System.Web
+		$proxyUri = "http://$( [System.Web.HttpUtility]::UrlEncode($proxyUsername) ):$( [System.Web.HttpUtility]::UrlEncode($proxyPassword) )@$( $proxyHost ):$( $proxyPort )"
```
Comment (johnsimons) [inline source/Calamari.Shared/Integration/Scripting/WindowsPowerShell/Bootstrap.ps1]: Shouldn't it be https for $env:HTTPS_PROXY ?

**Comment by johnsimons** [comment]
@benPearce1 I am still confused when user selects no proxy ? With this PR aren't we defaulting to system proxy ?
image

**Comment by johnsimons** [comment]
@benPearce1 I haven't verified this but I think if a user wants to use no proxy you only set the env NO_PROXY, you don't set the others, and the same for the opposite

File: source/Calamari.Shared/Integration/Scripting/WindowsPowerShell/Bootstrap.ps1 (hunk 3 above)
Comment (benPearce1) [reply source/Calamari.Shared/Integration/Scripting/WindowsPowerShell/Bootstrap.ps1]: This setting is consistent with Octopus Server and Tentacle https://github.com/OctopusDeploy/OctopusShared/blob/4d87e1cd0d6428b40c832523f387b23edda0885f/source/Octopus.Shared/Configuration/ProxyConfigParser.cs#L71

**Pull request merged**